ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/network/DefaultP2PNetwork.java

```diff
@@ -587,43 +583,44 @@ public void start() {
     };
   }
 
-  private synchronized void checkCurrentConnections() {
+  private void handlePermissionsUpdate(
+      final boolean permissionsRestricted, final Optional<List<Peer>> peers) {
+    if (!permissionsRestricted || !rlpxPermissions.isPresent()) {
+      // Nothing to do
+      return;
+    }
+
+    final PeerRlpxPermissions permissions = rlpxPermissions.get();
+    if (peers.isPresent()) {
+      peers.get().stream()
+          .filter(p -> !permissions.allowOngoingConnection(p))
+          .map(Peer::getId)
+          .map(connections::getConnectionForPeer)
+          .filter(Optional::isPresent)
+          .map(Optional::get)
+          .forEach(conn -> conn.disconnect(DisconnectReason.REQUESTED));
+    } else {
+      checkAllConnections();
+    }
+  }
+
+  private void checkAllConnections() {
+    if (!rlpxPermissions.isPresent()) {
+      return;
```

Comment:
This branch has no coverage.  Is this a case we expect to see in normal operations?  Would we ever realistically not have rlpxPermissions?

Reply:
The fact that this permissions object is wrapped in an `Optional` is unfortunate, and hopefully that will be going away in an upcoming PR.  I had tried initializing it in the constructor of `DefaultP2PNetwork` as a final unwrapped type that gets access to `DefaultP2PNetwork.getLocalEnode` via a `Supplier` argument in the `PeerRlpxPermissions` constructor.  But that caused issues with tests that needed to mock the `EnodeURL` returned by `DefaultP2pNetwork.getLocalEnode`.  For now, `rlpxPermissons` is created after the node starts and has a definite local enode.  

In this particular spot, this code will run when permissions are updated, which could happen before the node starts up.  Although, looks like I can simplify this a little by passing in the unwrapped permissions object - I'll push up that change.